File: citizensdk/android/src/main/kotlin/org/citizen/sdk/CitizenSdkFlutterWalletFlow.kt
package org.citizen.sdk

import androidx.fragment.app.FragmentActivity
import org.citizen.sdk.ui.CitizenSdkWalletFlowContract
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

/** One-shot ownership registry whose callback may win before handle binding. */
internal class CitizenSdkFlutterOneShotRegistry<K : Any, H : Any>(
    private val cancelHandle: (H) -> Unit,
) {
    internal class Entry<H : Any> {
        val handle = AtomicReference<H?>()
        val finished = AtomicBoolean(false)
        val cancelRequested = AtomicBoolean(false)
    }

    private val entries = ConcurrentHashMap<K, Entry<H>>()

    fun reserve(key: K): Entry<H>? = Entry<H>().takeIf { entries.putIfAbsent(key, it) == null }

    fun bind(entry: Entry<H>, handle: H) {
        check(entry.handle.compareAndSet(null, handle))
        if (entry.cancelRequested.get()) cancelHandle(handle)
    }

    fun finish(key: K, entry: Entry<H>): Boolean {
        if (!entry.finished.compareAndSet(false, true)) return false
        return entries.remove(key, entry)
    }

    fun cancelWhere(predicate: (K) -> Boolean) {
        entries.entries.filter { predicate(it.key) }.forEach { (_, entry) ->
            if (entry.cancelRequested.compareAndSet(false, true)) {
                entry.handle.get()?.let(cancelHandle)
            }
        }
    }

    internal fun sizeForTest(): Int = entries.size
}

/**
 * Secret-free Flutter projection of the SDK-owned wallet Activity.
 *
 * Flutter selects only the operation, word count, or public derivation
 * indices. Recovery phrases and passwords are entered/rendered by the
 * non-exported FLAG_SECURE Activity and never enter this object.
 */
internal class CitizenSdkFlutterWalletFlow {
    private data class Key(val sessionId: String, val requestSequence: Long)
    private val active = CitizenSdkFlutterOneShotRegistry<Key, () -> Unit> {
        it()
    }

    fun qr(sdk: CitizenSdk, activity: FragmentActivity?, request: CitizenSdkFlutterCodec.Request.Qr): CompletableFuture<CitizenQrDocument> {
        val host = activity ?: return failed(CitizenSdkException(CitizenSdkErrorCode.UNAVAILABLE, "QR 需要前台 FragmentActivity"))
        val key = Key(request.sessionId, request.requestSequence)
        val completion = CompletableFuture<CitizenQrDocument>()
        val owner = active.reserve(key) ?: return failed(CitizenSdkException(CitizenSdkErrorCode.CONFLICT, "QR 请求已存在"))
        try {
            val future: CompletableFuture<CitizenQrDocument>
            if (request.method == "signQrRequest") {
                val operation = sdk.signQrRequest(host, request.fields[0] as String)
                active.bind(owner) { operation.cancel(); Unit }
                future = operation.future.thenApply { it.document }
            } else {
                val operation = sdk.qrScan(host)
                active.bind(owner) { operation.cancel(); Unit }
                future = operation.future
            }
            future.whenComplete { value, error ->
                if (active.finish(key, owner)) {
                    if (error == null) completion.complete(value) else completion.completeExceptionally(error)
                }
            }
        } catch (error: Throwable) { if (active.finish(key, owner)) completion.completeExceptionally(error) }
        return completion
    }

    /** 私钥查看共享会话取消注册；完成值为空，不把内部 buffer 放入通道。 */
    fun viewAccountPrivateKey(sdk: CitizenSdk, activity: FragmentActivity?, request: CitizenSdkFlutterCodec.Request.Account): CompletableFuture<Unit> {
        val host = activity ?: return failed(CitizenSdkException(
            CitizenSdkErrorCode.UNAVAILABLE, "private key view requires a FragmentActivity",
        ))
        val key = Key(request.sessionId, request.requestSequence)
        val completion = CompletableFuture<Unit>()
        val owner = active.reserve(key) ?: return failed(CitizenSdkException(
            CitizenSdkErrorCode.CONFLICT, "wallet flow request already exists",
        ))
        try {
            val operation = sdk.viewAccountPrivateKey(host, request.accountId)
            active.bind(owner) { operation.cancel(); Unit }
            operation.future.whenComplete { _, failure ->
                if (active.finish(key, owner)) {
                    if (failure == null) completion.complete(Unit) else completion.completeExceptionally(failure)
                }
            }
        } catch (failure: Throwable) {
            if (active.finish(key, owner)) completion.completeExceptionally(failure)
        }
        return completion
    }

    fun launch(
        sdk: CitizenSdk,
        activity: FragmentActivity?,
        request: CitizenSdkFlutterCodec.Request.SessionRequest,
    ): CompletableFuture<CitizenWalletProfile?> {
        val host = activity ?: return failed(
            CitizenSdkException(
                CitizenSdkErrorCode.UNAVAILABLE,
                "CitizenSDK wallet UI requires a FragmentActivity",
            ),
        )
        val contract = contractRequest(request)
        val key = Key(request.sessionId, request.requestSequence)
        val completion = CompletableFuture<CitizenWalletProfile?>()
        val owner = active.reserve(key) ?: return failed(
            CitizenSdkException(CitizenSdkErrorCode.CONFLICT, "Wallet flow request already exists"),
        )
        try {
            val coordinator = sdk.launchWalletFlow(host, contract) { result ->
                if (!active.finish(key, owner)) return@launchWalletFlow
                when (result) {
                    is CitizenSdkWalletFlowContract.Result.Completed -> completion.complete(result.profile)
                    CitizenSdkWalletFlowContract.Result.Cancelled -> completion.completeExceptionally(
                        CitizenSdkException(CitizenSdkErrorCode.CANCELLED, "CitizenSDK wallet flow cancelled"),
                    )
                    is CitizenSdkWalletFlowContract.Result.Failed -> completion.completeExceptionally(result.error)
                }
            }
            active.bind(owner) { coordinator.cancel() }
        } catch (error: Throwable) {
            if (active.finish(key, owner)) completion.completeExceptionally(error)
        }
        return completion
    }

    /** Cancels every SDK-owned UI flow before supervised session destruction. */
    fun cancelSession(sessionId: String) {
        active.cancelWhere { it.sessionId == sessionId }
    }

    companion object {
        internal fun contractRequest(
            request: CitizenSdkFlutterCodec.Request.SessionRequest,
        ): CitizenSdkWalletFlowContract.Request = when (request) {
            is CitizenSdkFlutterCodec.Request.CreateWallet ->
                CitizenSdkWalletFlowContract.Request.Create(request.wordCount)
            is CitizenSdkFlutterCodec.Request.AddWalletAccounts ->
                CitizenSdkWalletFlowContract.Request.AddAccounts(request.indices)
            is CitizenSdkFlutterCodec.Request.Empty -> {
                require(request.method == "importWallet")
                CitizenSdkWalletFlowContract.Request.Import()
            }
            else -> throw CitizenSdkException(
                CitizenSdkErrorCode.INVALID_ARGUMENT,
                "Request is not an SDK-owned wallet flow",
            )
        }

        private fun <T> failed(error: Throwable): CompletableFuture<T> =
            CompletableFuture<T>().also { it.completeExceptionally(error) }
    }
}

File: citizensdk/darwin/Sources/CitizenSDKFlutter/CitizenSdkFlutterWalletFlow.swift
import CitizenSDK
import Foundation

#if os(iOS)
import UIKit
#elseif os(macOS)
import AppKit
#endif

/// Secret-free bridge to the SDK-owned Apple wallet interface.
@MainActor
internal final class CitizenSdkFlutterWalletFlow {
    private struct Key: Hashable { let session: String; let sequence: Int64 }
    private final class Entry {
        var handle: CitizenSDKWalletFlow?
        var operation: CitizenSDKOperation<Void>?
        var cancelQR: (() -> Void)?
        var cancelRequested = false
        var finished = false
    }
    private var active: [Key: Entry] = [:]

    /// 仅传递 Core 公开文档；相机、审阅和确认完全留在原生 SDK 窗口。
    func qr(sdk: CitizenSdk, request: CitizenSdkFlutterCodec.Request, signText: String?) async throws -> CitizenQRDocument {
        guard let session = request.sessionID, let sequence = request.sequence else {
            throw CitizenSDKError(.invalidArgument, "QR requires a session request")
        }
        let key = Key(session: session, sequence: sequence)
        guard active[key] == nil else { throw CitizenSDKError(.conflict, "QR request already exists") }
        let entry = Entry(); active[key] = entry
        defer { entry.finished = true; active.removeValue(forKey: key) }
        #if os(iOS)
        guard let presenter = Self.presenter() else { throw CitizenSDKError(.unavailable, "QR has no foreground presenter") }
        if let signText {
            let operation = try sdk.signQrRequest(from: presenter, text: signText)
            entry.cancelQR = { _ = try? operation.cancel() }
            if entry.cancelRequested { entry.cancelQR?() }
            return try await operation.value().document
        }
        let operation = try sdk.qrScan(from: presenter)
        #elseif os(macOS)
        guard let window = NSApp.keyWindow ?? NSApp.mainWindow else { throw CitizenSDKError(.unavailable, "QR has no host window") }
        if let signText {
            let operation = try sdk.signQrRequest(from: window, text: signText)
            entry.cancelQR = { _ = try? operation.cancel() }
            if entry.cancelRequested { entry.cancelQR?() }
            return try await operation.value().document
        }
        let operation = try sdk.qrScan(from: window)
        #endif
        entry.cancelQR = { _ = try? operation.cancel() }
        if entry.cancelRequested { entry.cancelQR?() }
        return try await operation.value()
    }

    /// 与已有钱包窗口共用取消注册，Flutter 只等真实 Void 终态，不接触显示缓冲。
    func viewAccountPrivateKey(sdk: CitizenSdk, request: CitizenSdkFlutterCodec.Request, accountID: Data) async throws {
        guard let session = request.sessionID, let sequence = request.sequence else {
            throw CitizenSDKError(.invalidArgument, "private key view requires a session request")
        }
        let key = Key(session: session, sequence: sequence)
        guard active[key] == nil else { throw CitizenSDKError(.conflict, "wallet flow request already exists") }
        let entry = Entry(); active[key] = entry
        defer { entry.finished = true; active.removeValue(forKey: key) }
        #if os(iOS)
        guard let presenter = Self.presenter() else { throw CitizenSDKError(.unavailable, "private key view has no foreground presenter") }
        let operation = try sdk.viewAccountPrivateKey(from: presenter, accountID: accountID)
        #elseif os(macOS)
        guard let window = NSApp.keyWindow ?? NSApp.mainWindow else { throw CitizenSDKError(.unavailable, "private key view has no host window") }
        let operation = try sdk.viewAccountPrivateKey(from: window, accountID: accountID)
        #endif
        entry.operation = operation
        if entry.cancelRequested { _ = try? operation.cancel() }
        try await operation.value()
    }

    func launch(sdk: CitizenSdk, request: CitizenSdkFlutterCodec.Request) async throws -> CitizenWalletProfile? {
        guard let session = request.sessionID, let sequence = request.sequence else {
            throw CitizenSDKError(.invalidArgument, "wallet flow requires a session request")
        }
        let contract = try Self.contract(for: request)
        let key = Key(session: session, sequence: sequence)
        guard active[key] == nil else { throw CitizenSDKError(.conflict, "wallet flow request already exists") }
        let entry = Entry(); active[key] = entry
        return try await withCheckedThrowingContinuation { continuation in
            do {
                #if os(iOS)
                guard let presenter = Self.presenter() else { throw CitizenSDKError(.unavailable, "CitizenSDK wallet UI has no foreground presenter") }
                let handle = try sdk.presentWalletFlow(from: presenter, request: contract) { [weak self, weak entry] result in
                    Task { @MainActor in self?.finish(key, entry: entry, result: result, continuation: continuation) }
                }
                #elseif os(macOS)
                guard let window = NSApp.keyWindow ?? NSApp.mainWindow else { throw CitizenSDKError(.unavailable, "CitizenSDK wallet UI has no host window") }
                let handle = try sdk.presentWalletFlow(from: window, request: contract) { [weak self, weak entry] result in
                    Task { @MainActor in self?.finish(key, entry: entry, result: result, continuation: continuation) }
                }
                #endif
                entry.handle = handle
                if entry.cancelRequested { handle.cancel() }
            } catch {
                active.removeValue(forKey: key)
                continuation.resume(throwing: error)
            }
        }
    }

    static func contract(for request: CitizenSdkFlutterCodec.Request) throws -> CitizenSDKWalletFlowRequest {
        switch request {
        case let .create(_, _, words): return .create(wordCount: words)
        case let .empty(method, _, _) where method == "importWallet": return .importWallet
        case let .addAccounts(_, _, indices): return .addAccounts(indices: indices)
        default: throw CitizenSDKError(.invalidArgument, "request is not an SDK-owned wallet flow")
        }
    }

    func cancelSession(_ session: String) {
        active.filter { $0.key.session == session }.forEach { _, entry in
            entry.cancelRequested = true
            entry.handle?.cancel()
            _ = try? entry.operation?.cancel()
            entry.cancelQR?()
        }
    }

    private func finish(_ key: Key, entry: Entry?, result: CitizenSDKWalletFlowResult,
                        continuation: CheckedContinuation<CitizenWalletProfile?, Error>) {
        guard let entry, active[key] === entry, !entry.finished else { return }
        entry.finished = true; active.removeValue(forKey: key)
        switch result {
        case let .completed(profile): continuation.resume(returning: profile)
        case .cancelled: continuation.resume(throwing: CitizenSDKError(.cancelled, "CitizenSDK wallet flow cancelled"))
        case let .failed(error): continuation.resume(throwing: error)
        @unknown default:
            continuation.resume(throwing: CitizenSDKError(.integrity, "CitizenSDK wallet flow returned an unknown result"))
        }
    }

    #if os(iOS)
    private static func presenter() -> UIViewController? {
        let root = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first(where: \.isKeyWindow)?.rootViewController
        var current = root
        while let presented = current?.presentedViewController { current = presented }
        if let navigation = current as? UINavigationController { return navigation.visibleViewController ?? navigation }
        if let tabs = current as? UITabBarController { return tabs.selectedViewController ?? tabs }
        return current
    }
    #endif
}
